site/_posts/2020-11-15-how-to-securely-implement-oauth-rails.md

```diff
@@ -0,0 +1,382 @@
+
+ ```ruby
+ Rails.application.routes.draw do
+   root to: 'welcome#index'
+   get '/oauth2-callback', to: 'oauth#oauth_callback'
+   get '/logout', to: 'oauth#logout'
+   get '/login', to: 'oauth#login'
+ end
+ ```
+
+#### The Welcome page
+Let's start off with a very basic welcome view that contains a link for a user to login to the application.
+This page will serve as the root page for our application.
+
+```ruby
+# app/views/welcome/index.html.erb
+
+<p>Welcome! Please <%= link_to 'sign in', controller: 'oauth', action: :login %> </p>
+```
+
+It should look something like this when rendered.
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/home.png" alt="The example Home screen for the Rails application." class="img-fluid" figure=false %}
+
+#### OauthController
+
+The "OauthController" will handle each of the remaining routes needed to complete the authorization code grant flow.
+First, we will setup our client using the [oauth2](https://github.com/oauth-xx/oauth2) gem. This provides us with a "REST" client wrapper for the OAuth 2.0 specification.
+
+```ruby
+class OauthController < ApplicationController
+  def initialize
+    @oauth_client = OAuth2::Client.new(Rails.configuration.x.oauth.client_id,
+                                       Rails.configuration.x.oauth.client_secret,
+                                       authorize_url: '/oauth2/authorize',
+                                       site: Rails.configuration.x.oauth.idp_url,
+                                       token_url: '/oauth2/token',
+                                       redirect_uri: Rails.configuration.x.oauth.redirect_uri)
+  end
+  # ...
+end
+```
+
+#### Login
+The login process beings when the user clicks on the "sign in" link we created.
+As seen in the sequence diagram above, we will want to direct this request to FusionAuth who is in charge of validating user credentials.
+In turn, the user will be presented with a default login page including username and password fields. 
+FusionAuth provides the flexibility to customize the style of this form via [themes](https://fusionauth.io/docs/v1/tech/themes/) but for this tutorial, we will stick with the default. 
+
+The following method uses the "OAuth2" client to construct the "authorize_url" and redirects the request accordingly. 
+```ruby
+def login
+  redirect_to @oauth_client.auth_code.authorize_url
+end
+```
+
+A successful response will yield the login page:
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/fa-login.png" alt="The FusionAuth login page." class="img-fluid" figure=false %}
+
+#### The OAuth callback
+
+Upon successful authentication, FusionAuth redirects to the Rails app along with the "authorization code" using the "redirect_uri".
+This value tells FusionAuth where to redirect upon successful authentication of which corresponds with our `oauth_callback` route.
```

Comment:
"successful authentication of which corresponds "
is awk? maybe 

"uccessful authentication which corresponds "